src/lib-ssl-iostream/iostream-openssl.c

```diff
@@ -811,6 +812,86 @@ openssl_iostream_get_peer_name(struct ssl_iostream *ssl_io)
 #endif
 	i_assert(x509 != NULL);
 
+	if (ssl_io->username_cea) {
+
+		ASN1_INTEGER *serialNumber;
+		X509_NAME *issuer;
+		BIGNUM *bn;
+		char *decimal;
+		BIO *bio;
+
+		bio = BIO_new(BIO_s_mem());
+		if (!bio) {
+			name = NULL;
+			goto end;
+		}
+
+		BIO_puts(bio, "{ serialNumber ");
+
+		serialNumber = X509_get_serialNumber(x509);
+		if (!serialNumber) {
+			BIO_free(bio);
+			name = NULL;
+			goto end;
+		}
+
+		bn = ASN1_INTEGER_to_BN(serialNumber, NULL);
+		if (!bn) {
+			BIO_free(bio);
+			name = NULL;
+			goto end;
+		}
+
+		decimal = BN_bn2dec(bn);
+		if (!decimal) {
+			BIO_free(bio);
+			BN_free(bn);
+			name = NULL;
+			goto end;
+		}
+
+		BIO_puts(bio, decimal);
+
+		OPENSSL_free(decimal);
+		BN_free(bn);
+
+		BIO_puts(bio, ", issuer rdnSequence:\"");
+
+		issuer = X509_get_issuer_name(x509);
```

Comment:
This seems like it should be escaping `\` and `"` in case they happen to exist in the issuer name. str_escape() would do that. And seems like this would be a bit simpler if BIO was used only for the X509_NAME_print_ex() and the rest used either t_strdup_printf() or string_t.

Reply:
> This seems like it should be escaping \ and " in case they happen to exist in the issuer name. str_escape() would do that

Let me check this.

> And seems like this would be a bit simpler if BIO was used only for the X509_NAME_print_ex() and the rest used either t_strdup_printf() or string_t.

Felt a bit weird to mix them up, but can do.

Reply:
> This seems like it should be escaping \ and " in case they happen to exist in the issuer name. 

Checked - escaping is unnecessary (and would break the string), as the rfc2253 already escapes the issuer name as described below:

https://datatracker.ietf.org/doc/html/rfc2253#section-2.4

Reply:
> And seems like this would be a bit simpler if BIO was used only for the X509_NAME_print_ex() and the rest used either t_strdup_printf() or string_t.

Looking at the code, using a BIO is the simplest way.

If I had to try and convert it to t_strdup_printf(), we would still have the same extraction of the fields, then we would need to t_malloc0() the BIO, then assemble the string, then get rid of the temporary t_malloc0.

Letting openssl do all the work is way cleaner.

Reply:
> Checked - escaping is unnecessary (and would break the string), as the rfc2253 already escapes the issuer name as described below:

I think the certificates could still be evil and have `"` character without escaping it?

Reply:
> I think the certificates could still be evil and have " character without escaping it? So some sanity check for the string would still be useful.

That's not possible in this case.

RFC4523 appendix A.1 describes a string that consists of the serial number (in a safe format), concatenated with the issuer DN encoded as per rfc2253 (in a safe format).

If the issuer wasn't safely escaped already, someone could create a malicious certificate that contained the target desired issuer with a 0x00 character in it, and then fake the identity of another certificate. This is prevented because rfc2253 mandates that the distinguished name be escaped.

Quoting from [rfc2253](https://datatracker.ietf.org/doc/html/rfc2253#section-2.4) to show we are escaped already:

>  If the UTF-8 string does not have any of the following characters
>    which need escaping, then that string can be used as the string
>    representation of the value.
> 
>     o   a space or "#" character occurring at the beginning of the
>         string
> 
>     o   a space character occurring at the end of the string
> 
>     o   one of the characters ",", "+", """, "\", "<", ">" or ";"
> 
>    Implementations MAY escape other characters.
> 
>    If a character to be escaped is one of the list shown above, then it
>    is prefixed by a backslash ('\' ASCII 92).
> 
>    Otherwise the character to be escaped is replaced by a backslash and
>    two hex digits, which form a single byte in the code of the
>    character.
> 

Any attempt to modify the CEA string outside that described by the RFC above means the string will not match with or interoperate with other software.